sql: prepared statements cache parsed query

jordanlewis opened the pull request:
Previously, prepared statements contained only the text of their query, causing them to be re-parsed at execution time.

Now, they contain a parsed query.

This improves throughput of `kv -read-percent=100 -concurrency=1` by about 15% on a local single-node cluster.

Updates #14927.

jordanlewis committed:
sql: remove placeholders using a visitor

Previously, placeholder nodes were completely removed during
type-checking when a value was available for them and replaced by their
value. This mutated the expression tree and made it unsuitable for
re-use.

Now, the expression tree is "modified" using a visitor, which ensures
that the original copy of the expression tree is left unchanged.

This new behavior preserves the original parsed expression tree even
after evaluation, which sets the stage for caching parse results.

jordanlewis committed:
sql: prepared statements cache parsed statement

Previously, prepared statements contained only the text of their query,
causing them to be re-parsed at execution time.

Now, they contain a parsed query.

This improves throughput of kv -read-percent=100 -concurrency=1 by about
15% on a local single-node cluster.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/15639)
<!-- Reviewable:end -->

petermattis commented:
Nice! Tagging with 1.1 as it is too late for this to go into 1.0. I didn't look at the change closely, but wanted to point out #14932 which also tackles this (though unsuccessfully).

tamird commented:
Wow, that's an impressive improvement!

Can you help me understand why the first commit is necessary? Wouldn't it be valid to cache the parsed result with the placeholders substituted out?

---

Reviewed 3 of 3 files at r1.
Review status: 3 of 6 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-:-KjDfd_biDfc2Ldr2wmd:b-eggy74)*
<!-- Sent from Reviewable.io -->

knz commented:
Nice work indeed.

The first commit is necessary in that Eval should not panic on placeholders. But the code can be simplified, see my comments below.

---

Reviewed 3 of 3 files at r1, 3 of 3 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/sql/executor.go, line 507 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjDhIM9Zx_ncgOoT0UG:-KjDhIM9Zx_ncgOoT0UH:b-t8d55z) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cdffd8b1945d4d89c22bcceed3b2e5916343d9f/pkg/sql/executor.go#L507)):*
> ```Go
> 
> // ExecuteStatementsParsed executes the given statement and returns a response.
> func (e *Executor) ExecuteStatementParsed(
> ```

1) `ExecutePreparedStatement` is closer to the truth 
2) since this is only executing prepared statements, what about giving the prepared statement key as argument and move the prepared statement lookup to here.

---

*[pkg/sql/executor.go, line 513 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjDgJ4dCCvr8LOlbxL7:-KjDgJ4dCCvr8LOlbxL8:bob8csc) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cdffd8b1945d4d89c22bcceed3b2e5916343d9f/pkg/sql/executor.go#L513)):*
> ```Go
> 	session.phaseTimes[sessionStartBatch] = timeutil.Now()
> 
> 	defer func() {
> ```

Make this deferred function a common helper. Possibly take inspiration from #15358 (still planning to merge that one eventually) 

---

*[pkg/sql/parser/eval.go, line 2818 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjDfrA6mD3evHrN61yW:-KjDfrA6mD3evHrN61yX:biwf0it) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bd6ba4917693ac81827469a073bb678a711898f/pkg/sql/parser/eval.go#L2818)):*
> ```Go
> // Eval implements the TypedExpr interface.
> func (node *Placeholder) Eval(ctx *EvalContext) (Datum, error) {
> 	return node.Value.Eval(ctx)
> ```

`return ctx.Placeholders.Value(node.Name)`

---

*[pkg/sql/parser/expr.go, line 616 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjDfvc48xQ0-214QRqs:-KjDfvc48xQ0-214QRqt:ba8jqxq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bd6ba4917693ac81827469a073bb678a711898f/pkg/sql/parser/expr.go#L616)):*
> ```Go
> type Placeholder struct {
> 	Name  string
> 	Value TypedExpr
> ```

remove this once `Eval` uses `ctx` to retrieve the value.

---

*[pkg/sql/parser/expr.go, line 633 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjDfzY4LPSkFbMZDweY:-KjDfzY4LPSkFbMZDweZ:b-dksx3l) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bd6ba4917693ac81827469a073bb678a711898f/pkg/sql/parser/expr.go#L633)):*
> ```Go
> 	buf.WriteByte('$')
> 	buf.WriteString(node.Name)
> 	if node.Value != nil {
> ```

Not a fan of this. I don't see any benefits.

---

*[pkg/sql/parser/type_check.go, line 738 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjDg2DiuTRIcFVHfnse:-KjDg2DjgtqSL6TESg5s:b-q64pkt) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bd6ba4917693ac81827469a073bb678a711898f/pkg/sql/parser/type_check.go#L738)):*
> ```Go
> 	// If there is a value known already, immediately substitute with it.
> 	if v, ok := ctx.Placeholders.Value(expr.Name); ok {
> 		expr.Value = v
> ```

See above.

---

*[pkg/sql/pgwire/v3.go, line 596 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjDgkpc0NxNeh8JMFHm:-KjDgkpc0NxNeh8JMFHn:bqdeh5a) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cdffd8b1945d4d89c22bcceed3b2e5916343d9f/pkg/sql/pgwire/v3.go#L596)):*
> ```Go
> // type should return NoData.
> func canSendNoData(stmt parser.Statement) bool {
> 	return stmt == nil || stmt.StatementType() != parser.Rows
> ```

Please revise this and above -- unfortunately Go also has empty but non-nil arrays. So you could have a parser.Statement reference that is non-nill but still should return `canSendNoData` true.

I think we should really have a `StatementList` reference in PreparedStatement, not `Statement`, and use that here too with `len(stmt) == 0` instead of `stmt == nil`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-:-KjDhTPOIQmDtvaCnFUF:b-stuci1)*
<!-- Sent from Reviewable.io -->

knz commented:
@tamird in SQL you can execute a prepared statement multiple times with different placeholder values. 

---

Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-:-KjDhgBYn1bSqfvy6E2R:b-6kx90y)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Whoops, thanks for the pointer @petermattis as I totally missed your attempt.

@tamird without that patch, the following sequence of commands would return `3` both times, since executing a statement with placeholder values available actually modifies the parsed query:

```
PREPARE q AS SELECT $1;
EXECUTE q(3);
EXECUTE q(4);
```

---

Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-:-KjDgwAcNxhwOBTt4XQH:b3hpdcf)*
<!-- Sent from Reviewable.io -->

tamird commented:
Ah, of course. Thanks @knz @jordanlewis!

---

Reviewed 1 of 3 files at r2.
Review status: all files reviewed at latest revision, 7 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-:-KjDi02HEV33O75oFmx_:b-1hoytn)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
TFTR @knz!

---

Review status: all files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/sql/executor.go, line 507 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjDhIM9Zx_ncgOoT0UG:-KjEJfYGJw47kzkGQU7T:b-p4bm4e) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cdffd8b1945d4d89c22bcceed3b2e5916343d9f/pkg/sql/executor.go#L507)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

1) `ExecutePreparedStatement` is closer to the truth 
2) since this is only executing prepared statements, what about giving the prepared statement key as argument and move the prepared statement lookup to here.
</blockquote></details>

Renamed. I think giving the key isn't quite right, because each of the calling methods do different things with the prepared statement.

---

*[pkg/sql/executor.go, line 513 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjDgJ4dCCvr8LOlbxL7:-KjEKdgSKxwgVNonkQJc:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cdffd8b1945d4d89c22bcceed3b2e5916343d9f/pkg/sql/executor.go#L513)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Make this deferred function a common helper. Possibly take inspiration from #15358 (still planning to merge that one eventually) 
</blockquote></details>

Done.

---

*[pkg/sql/parser/eval.go, line 2818 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjDfrA6mD3evHrN61yW:-KjEPOVH5rolVob2HX1b:b16pj1a) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bd6ba4917693ac81827469a073bb678a711898f/pkg/sql/parser/eval.go#L2818)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

`return ctx.Placeholders.Value(node.Name)`
</blockquote></details>

I think the Eval is necessary because placeholders can contain Expressions now and not only Datums.

Unfortunately, this is also the wrong kind of `ctx`... the `SemaContext` has the placeholders, not the `EvalContext`. I don't think there's a way to get from one to another right now... thoughts?

I agree that your solution is cleaner - there's no point in mutating the expression tree to include the placeholder value.

---

*[pkg/sql/parser/expr.go, line 633 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjDfzY4LPSkFbMZDweY:-KjEbkoqcr582ikvDjg_:bss4a2b) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bd6ba4917693ac81827469a073bb678a711898f/pkg/sql/parser/expr.go#L633)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Not a fan of this. I don't see any benefits.
</blockquote></details>

Removed.

---

*[pkg/sql/pgwire/v3.go, line 596 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjDgkpc0NxNeh8JMFHm:-KjErdB_OVsDLUbO530w:b-7c2fp2) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cdffd8b1945d4d89c22bcceed3b2e5916343d9f/pkg/sql/pgwire/v3.go#L596)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

Please revise this and above -- unfortunately Go also has empty but non-nil arrays. So you could have a parser.Statement reference that is non-nill but still should return `canSendNoData` true.

I think we should really have a `StatementList` reference in PreparedStatement, not `Statement`, and use that here too with `len(stmt) == 0` instead of `stmt == nil`.
</blockquote></details>

I think it's misleading to have a `StatementList` within `PreparedStatement`, though, because it should only have size 1 maximum. And I can't find a case where `stmt` here is non-nil but also should return no data - when we create a PreparedStatement in the first place, if there's no statements to process, we return a PreparedStatement with a `nil` Statement inside... please show me what I'm missing?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-:-KjEJJSlrxGkUYFRPLoT:b-qnxvkx)*
<!-- Sent from Reviewable.io -->

knz commented:
Reviewed 1 of 4 files at r3, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/parser/eval.go, line 2818 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjDfrA6mD3evHrN61yW:-KjF7-hKaNFzhcLSQOd3:bk7nmfq) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bd6ba4917693ac81827469a073bb678a711898f/pkg/sql/parser/eval.go#L2818)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I think the Eval is necessary because placeholders can contain Expressions now and not only Datums.

Unfortunately, this is also the wrong kind of `ctx`... the `SemaContext` has the placeholders, not the `EvalContext`. I don't think there's a way to get from one to another right now... thoughts?

I agree that your solution is cleaner - there's no point in mutating the expression tree to include the placeholder value.
</blockquote></details>

so indeed `return ctx.xxx.Get(node.Name).Eval(ctx)`
The placeholder values need to be plumbed through the EvalContext, indeed.

Meanwhile, whether you make this change or not, you will also need to check that placeholder values are properly propagated to distsql processors. I suspect they are not, until you extend the distsql protobufs.

---

*[pkg/sql/pgwire/v3.go, line 596 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjDgkpc0NxNeh8JMFHm:-KjF7XnbVyDRxFAUfWGC:b-3dyt6t) ([raw file](https://github.com/cockroachdb/cockroach/blob/1cdffd8b1945d4d89c22bcceed3b2e5916343d9f/pkg/sql/pgwire/v3.go#L596)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I think it's misleading to have a `StatementList` within `PreparedStatement`, though, because it should only have size 1 maximum. And I can't find a case where `stmt` here is non-nil but also should return no data - when we create a PreparedStatement in the first place, if there's no statements to process, we return a PreparedStatement with a `nil` Statement inside... please show me what I'm missing?
</blockquote></details>

Ok, I can see how the current code limits this. I'll let it sleep until we start supporting stored procedures.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639)*
<!-- Sent from Reviewable.io -->

jordanlewis committed:
sql: allow array datum-to-col-type conversion

jordanlewis commented:
Tests should pass now, and DistSQL should work correctly.

I discussed the best approach for marshalling placeholders over DistSQL with @raduberinde offline, and we came to the conclusion that it's easiest to replace placeholder values at serialization time rather than trying to send the placeholders to all relevant processors. That's what I've done here.

---

Review status: 0 of 9 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[pkg/sql/parser/eval.go, line 2818 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjDfrA6mD3evHrN61yW:-KjPZsyuGdvCFs0ZN9ej:b-dpb2cu) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bd6ba4917693ac81827469a073bb678a711898f/pkg/sql/parser/eval.go#L2818)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

so indeed `return ctx.xxx.Get(node.Name).Eval(ctx)`
The placeholder values need to be plumbed through the EvalContext, indeed.

Meanwhile, whether you make this change or not, you will also need to check that placeholder values are properly propagated to distsql processors. I suspect they are not, until you extend the distsql protobufs.
</blockquote></details>

I'm afraid that doing this correctly will be quite onerous, as many more values will need to be plumbed deep within DistSQL in all sorts of unexpected places (every time `MakeExpression` is called), so for the time being I've left the value expression within the `Placeholder` node.

I agree that it's a little messy to store the value on the `Placeholder`, not to mention dangerous since it is not cleared after execution, which exposes us to seeing old placeholder values if there are bugs in our code.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-:-KjPZh-WZp3vfEo_8lpX:b-ux108w)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Whoops, I spoke too soon about the tests.

andreimatei commented:
Review status: 0 of 9 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/parser/eval.go, line 2818 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjDfrA6mD3evHrN61yW:-KjPaWzhX5Wv800Zt-vY:b-idi1nm) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bd6ba4917693ac81827469a073bb678a711898f/pkg/sql/parser/eval.go#L2818)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I'm afraid that doing this correctly will be quite onerous, as many more values will need to be plumbed deep within DistSQL in all sorts of unexpected places (every time `MakeExpression` is called), so for the time being I've left the value expression within the `Placeholder` node.

I agree that it's a little messy to store the value on the `Placeholder`, not to mention dangerous since it is not cleared after execution, which exposes us to seeing old placeholder values if there are bugs in our code.
</blockquote></details>

given what you said in the top-level discussion about how DistSQL will not need to deal with placeholders, can we now get rid of `Placeholder.Value`, and do things along the lines Rafa's suggesting?

Better yet, I think it'd be nice if we had a pass (a Visitor) that does the substitution like the TypeChecking used to. Since supposedly we already have something like that for the distsql rendering? Then we wouldn't need to complicate any EvalContexts.
Does that make sense?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Review status: 0 of 9 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/parser/eval.go, line 2818 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjDfrA6mD3evHrN61yW:-KjTIfjwzHbfx4g7bzdz:bqlit5n) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bd6ba4917693ac81827469a073bb678a711898f/pkg/sql/parser/eval.go#L2818)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

given what you said in the top-level discussion about how DistSQL will not need to deal with placeholders, can we now get rid of `Placeholder.Value`, and do things along the lines Rafa's suggesting?

Better yet, I think it'd be nice if we had a pass (a Visitor) that does the substitution like the TypeChecking used to. Since supposedly we already have something like that for the distsql rendering? Then we wouldn't need to complicate any EvalContexts.
Does that make sense?
</blockquote></details>

We can't naively get rid of `Placeholder.Value` without plumbing the placeholders map into everywhere `MakeExpression` is called, which is what I mentioned above

Your suggestion to use a Visitor is another possible solution but does introduce some questions. If we were to use a Visitor, which is basically what the old code did to replace `Placeholder`s with their values, we'd again break the assumption that the parse tree stays unmodified after its initial creation. Alternatively, we could make a deep copy of the parse tree right before executing a prepared statement. Perhaps that is the correct solution here?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 0 of 9 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/parser/eval.go, line 2818 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjDfrA6mD3evHrN61yW:-KjTOy_sXpjJxca8YPEg:b1hr1eg) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bd6ba4917693ac81827469a073bb678a711898f/pkg/sql/parser/eval.go#L2818)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

We can't naively get rid of `Placeholder.Value` without plumbing the placeholders map into everywhere `MakeExpression` is called, which is what I mentioned above

Your suggestion to use a Visitor is another possible solution but does introduce some questions. If we were to use a Visitor, which is basically what the old code did to replace `Placeholder`s with their values, we'd again break the assumption that the parse tree stays unmodified after its initial creation. Alternatively, we could make a deep copy of the parse tree right before executing a prepared statement. Perhaps that is the correct solution here?
</blockquote></details>

FYI, I experimented with making a deep copy of the parse tree (using reflection to make the copy). That was a non-starter from a performance perspective. Might want to verify that result, and not using reflection could change the numbers. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Review status: 0 of 9 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[pkg/sql/parser/eval.go, line 2818 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjDfrA6mD3evHrN61yW:-KjThMeVCH45XcdrWkAE:b-aiub3b) ([raw file](https://github.com/cockroachdb/cockroach/blob/1bd6ba4917693ac81827469a073bb678a711898f/pkg/sql/parser/eval.go#L2818)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

FYI, I experimented with making a deep copy of the parse tree (using reflection to make the copy). That was a non-starter from a performance perspective. Might want to verify that result, and not using reflection could change the numbers. 
</blockquote></details>

@jordanlewis Visitors don't normally change the expression in-place. When you return a new expression from`VisitPre/VisitPost`, the parent nodes are copied as necessary so that the old tree remains valid. See `WalkExpr`.

I implemented this "immutable nodes model" as an alternative to implementing deep copy (which was even more tedious to do). There are various things that do mutate nodes in-place so we're not strictly adhering to the model, but those changes are (so far) harmless.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Ah I see. Thanks. I was going to do a comparison between the mutable method I'm using currently and a new `WalkExpr` implementation, when I noticed that this patch now tanks the performance vs master.

I think the main change between now and when I first checked the performance is that DistSQL is always on, correct? I'll need to figure out why this is making things worse now.

RaduBerinde commented:
The simple query done by `kv` shouldn't be running through DistSQL in auto mode. We even have a test for that (explain_distsql:25).

jordanlewis commented:
I see. The trouble was that index selection expects placeholders to have been completely replaced by `Datum` values. That test is unfortunately insufficient because it doesn't try to use prepared queries, and we can't write an `EXPLAIN` test against prepared queries yet. I filed #15743 to add support for that.

I've redone this patch to use a visitor to replace the placeholders right before type checking. This is considerably simpler and doesn't require messing with the expression serialization code. PTAL.

Here are the new performance numbers. I don't see the 15% speedup anymore, now it's more like 5%. This is at least partially caused by the extra copies introduced by the visitor, but there's also considerable variance when testing this stuff locally.

Before:
```
[21:56]% ./kv -read-percent=100 -concurrency=1 -duration=10s
_elapsed___errors__ops/sec(inst)___ops/sec(cum)__p95(ms)__p99(ms)_pMax(ms)
      1s        0         2790.2         2790.1      0.5      1.4      3.3
      2s        0         3121.9         2955.6      0.4      0.7      4.5
      3s        0         3105.3         3005.5      0.4      0.7      4.1
      4s        0         3030.1         3011.6      0.5      0.7      2.8
      5s        0         3117.8         3032.9      0.4      0.7      3.1
      6s        0         3122.4         3047.8      0.4      0.7      3.5
      7s        0         3124.7         3058.8      0.4      0.7      4.2
      8s        0         3035.6         3055.9      0.4      0.7      3.7
      9s        0         3049.5         3055.2      0.4      0.6      5.8
     10s        0         3101.3         3059.8      0.4      0.6      5.0

_elapsed___errors____________ops___ops/sec(cum)__p95(ms)__p99(ms)_pMax(ms)_____seq(begin/end)
   10.0s        0          30608         3059.8      0.4      0.7      5.8              0 / 0

BenchmarkBlocks	   30608	    326824.5 ns/op
```

After:
```
[21:43:130]% ./kv -read-percent=100 -concurrency=1 -duration=10s
_elapsed___errors__ops/sec(inst)___ops/sec(cum)__p95(ms)__p99(ms)_pMax(ms)
      1s        0         3016.9         3016.8      0.5      0.7      3.4
      2s        0         3333.4         3174.8      0.4      0.6      2.4
      3s        0         3298.2         3215.9      0.4      0.6      3.1
      4s        0         3278.9         3231.7      0.4      0.6      2.6
      5s        0         3276.6         3240.6      0.4      0.6      3.5
      6s        0         3259.5         3243.8      0.4      0.7      3.3
      7s        0         3297.6         3251.5      0.4      0.7      2.9
      8s        0         3230.3         3248.8      0.4      0.6      3.9
      9s        0         3113.6         3233.8      0.5      0.7      5.5
     10s        0         3300.2         3240.4      0.4      0.6      3.8

_elapsed___errors____________ops___ops/sec(cum)__p95(ms)__p99(ms)_pMax(ms)_____seq(begin/end)
   10.0s        0          32419         3240.5      0.4      0.7      5.5              0 / 0

BenchmarkBlocks	   32419	    308592.1 ns/op
```

knz commented:
Your solution albeit functionally correct is ... potentially multiplying the number of allocations by the size of the ASTs.

I still think you can do better. Like this:

- do like you did in one of your earlier attempts: have TypeCheck keep the placeholders, and have Eval look the value up from the context
- use a callback in `FmtParsable`, used by distsql to "serialize" expressions, to replace the placeholders by their value datums during the AST to string conversion. This way the distsql processors don't see any placeholder any more.

This saves you all the allocations and many tree traversals.

---

Reviewed 6 of 10 files at r5, 5 of 5 files at r6, 1 of 1 files at r7, 3 of 3 files at r8.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/sql/prepare.go, line 101 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjZTZbcp7VFbMdXs3zP:-KjZTZbcp7VFbMdXs3zQ:b-8rqb8l) ([raw file](https://github.com/cockroachdb/cockroach/blob/35af9e2e07a518a6dc2deafa2ec21794f6e3a99b/pkg/sql/prepare.go#L101)):*
> ```Go
> 	}
> 
> 	sz := int64(uintptr(len(name)) + unsafe.Sizeof(*stmt))
> ```

This change is incorrect and removing the comment is misleading.
The correct code here should measure the memory usage of the AST that's being retained. I understand you can't or don't want to do this here, but an outrageously large query should show up in the memory monitoring still. At least you could keep the length of the query string as an estimator of the memory usage -- the reasoning being that the ASCII encoding in SQL is not a very large constant factor away from the AST structural representation. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-:-KjZUojb3RiKKWpeYudw:bou4g27)*
<!-- Sent from Reviewable.io -->

jordanlewis committed:
sql: export parser.Parse

Previously, internal users that wanted to use a parser to parse a sql
string containing more than one statement needed to declare a
`parser.Parser` and call its `Parse` method.

Now, `parser.Parse` is exported which does the exact same thing.

jordanlewis commented:
As discussed offline, the approach you suggest above is what I tried in the previous iteration of this PR (mod keeping the placeholder values somewhere besides the `Placeholder` nodes), and I abandoned it due to the additional complexity of properly teaching the index selection and analyze code about the possible existence of Placeholders.

I am planning to do that work in a separate PR, since it's a bit beyond the scope of this one. I've opened #15792 to track the remaining work.

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, some commit checks failed.

---

*[pkg/sql/prepare.go, line 101 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjZTZbcp7VFbMdXs3zP:-KjeEtAUL6zstLjz-lbG:b-8k5cnb) ([raw file](https://github.com/cockroachdb/cockroach/blob/35af9e2e07a518a6dc2deafa2ec21794f6e3a99b/pkg/sql/prepare.go#L101)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

This change is incorrect and removing the comment is misleading.
The correct code here should measure the memory usage of the AST that's being retained. I understand you can't or don't want to do this here, but an outrageously large query should show up in the memory monitoring still. At least you could keep the length of the query string as an estimator of the memory usage -- the reasoning being that the ASCII encoding in SQL is not a very large constant factor away from the AST structural representation. 
</blockquote></details>

Good point. I've restored the old behavior.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-:-KjeF-hrVpXgkisuBm3R:b-p2hq2h)*
<!-- Sent from Reviewable.io -->

knz commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 4 of 8 files at r9, 1 of 1 files at r10, 3 of 4 files at r11, 3 of 3 files at r12.
Review status: all files reviewed at latest revision, 3 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-:-Kjh2NTCwCWalN5gFdIL:bnfp4nl)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Some post-merge comments. Sorry for the late review. Nothing major though.

One high-level thing I'd like to understand: does this change make us keep more state in a session than we were before for statements that are not necessarily going to be repeated? I know some drivers prepare all the statements before executing them. Are we going to keep state in that case (and if yes, are we now gonna keep more state)? Or do we somehow get an indication (maybe through the name) about what's really going to be reused and what's prepared for other reasons?

---

Review status: all files reviewed at latest revision, 19 unresolved discussions, all commit checks successful.

---

*[pkg/sql/executor.go, line 404 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiVEb_h6kALjcXjf9N:-KjiVEb_h6kALjcXjf9O:be6pbv5) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3023d4f47d400559e158c4b3ea9aa9477af319e/pkg/sql/executor.go#L404)):*
> ```Go
> // nil if there are no results).
> func (e *Executor) Prepare(
> 	stmts parser.StatementList, session *Session, pinfo parser.PlaceholderTypes,
> ```

I spent a while scratching my head about whether this is supposed to be a single statement or it can also be multiple ones until I saw the `switch len(stmts)` below.
Mind documenting that it needs to be a single statement?
Better yet, I think it'd be better if this took `parser.Statement` and we'd move error checking to the callers.

---

*[pkg/sql/executor.go, line 482 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiMC6aLkQu3_7oSNVT:-KjiMC6aLkQu3_7oSNVU:b-9j0kg8) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3023d4f47d400559e158c4b3ea9aa9477af319e/pkg/sql/executor.go#L482)):*
> ```Go
> }
> 
> func logIfPanicking(ctx context.Context, sql string) {
> ```

nit: comment that this function needs to be directly `defer`red. Like, it can't be called from an anonymous deferred function.

---

*[pkg/sql/executor.go, line 508 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiO7mfkDv2vdNH2CU0:-KjiO7mfkDv2vdNH2CU1:b1rn0x2) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3023d4f47d400559e158c4b3ea9aa9477af319e/pkg/sql/executor.go#L508)):*
> ```Go
> 	session *Session, stmt *PreparedStatement, pinfo *parser.PlaceholderInfo,
> ) StatementResults {
> 	session.resetForBatch(e)
> ```

are you sure this `resetForBatch()` here is correct? I don't know exactly what the consequences of it are, but the name certainly indicates that it should be done per batch of statements.
It's also called from `Executor.Prepare`, which I'm guessing is also wrong now that we support `PREPARE` as a sql statement, and not just as a wire command that always runs outside of batches.

---

*[pkg/sql/executor.go, line 509 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiOtZFPzGoYLvT6ipn:-KjiOtZFPzGoYLvT6ipo:brgob4g) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3023d4f47d400559e158c4b3ea9aa9477af319e/pkg/sql/executor.go#L509)):*
> ```Go
> ) StatementResults {
> 	session.resetForBatch(e)
> 	session.phaseTimes[sessionStartBatch] = timeutil.Now()
> ```

same for this `phaseTimes` - is the "startBatch" correct?

---

*[pkg/sql/prepare.go, line 31 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiQys6zZs2jMlkJXoo:-KjiQys7XqdnZ5Cdwx_A:b-1cq3os) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3023d4f47d400559e158c4b3ea9aa9477af319e/pkg/sql/prepare.go#L31)):*
> ```Go
> // of arguments and results have been determined.
> type PreparedStatement struct {
> 	// Statement is the parsed, prepared SQL statement. It may be nil if the
> ```

Mind explaining (or just telling me) what an "empty" prepared statement represents? What does it return?

---

*[pkg/sql/prepare.go, line 73 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiSpKhnCGkKI32uF8M:-KjiSpKhnCGkKI32uF8N:b-fsepjq) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3023d4f47d400559e158c4b3ea9aa9477af319e/pkg/sql/prepare.go#L73)):*
> ```Go
> // in inferring placeholder types.
> //
> // ps.session.Ctx() is used as the logging context for the prepare operation.
> ```

I don't think "logging context" is a phrase we should be using. Contexts help with more than just logging.

Feel free to ignore this, but I don't think this comment is necessary. I suspect there's no real need for the `PreparedStatements` type to hold a reference to the session; it could probably be passed in everywhere. And then you probably wouldn't have felt the need to write it, as it's more common knowledge that sessions have a context (or at least something more magic that provides a context) for all their operations. Or if we keep the session ptr, I think a comment on the `PreparedStatements` type saying that an instance is a part of a `Session` would be better,

---

*[pkg/sql/prepare.go, line 74 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiYRLecbVds5ehVTIo:-KjiYRLecbVds5ehVTIp:b-ulkmz4) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3023d4f47d400559e158c4b3ea9aa9477af319e/pkg/sql/prepare.go#L74)):*
> ```Go
> //
> // ps.session.Ctx() is used as the logging context for the prepare operation.
> func (ps PreparedStatements) NewFromString(
> ```

Is this new method (which also comes with a duplicate comment) really necessary? Can't the caller do the parsing and call the other one?

---

*[pkg/sql/prepare.go, line 95 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiZHNdRmvohmDgS3Or:-KjiZHNdRmvohmDgS3Os:b7ay9sq) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3023d4f47d400559e158c4b3ea9aa9477af319e/pkg/sql/prepare.go#L95)):*
> ```Go
> 	name string,
> 	stmts parser.StatementList,
> 	queryLen int,
> ```

the `queryLen` argument seems very odd to me. We moved to a nice structured type for the `statement` (really, `StatementList`, but as I was saying above I think that's a mistake), but we pass this length separately. And also the query string length is pretty silly IMHO for the memory mgmt purposes for which it's used. I suggest adding a `Size()` method to a `parser.Statement` which takes into account more than just the query length, and getting rid of this argument.

---

*[pkg/sql/parser/placeholders.go, line 168 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiK_8hCYi6NJFAFiA1:-KjiK_8i__LTCEj3A63s:bgksil4) ([raw file](https://github.com/cockroachdb/cockroach/blob/4092ccef0e8746218e05658c9408a1cec97bc909/pkg/sql/parser/placeholders.go#L168)):*
> ```Go
> 	switch t := expr.(type) {
> 	case *Placeholder:
> 		if e, ok := v.placeholders.Value(t.Name); ok {
> ```

what if a value is not found? Should that be an error?
This probably deserves to be documented on the `placeholderVisitors` type definition.

---

*[pkg/sql/parser/placeholders.go, line 169 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiKNPKDiqVAPFyjzQ8:-KjiKNPKDiqVAPFyjzQ9:bpk8hez) ([raw file](https://github.com/cockroachdb/cockroach/blob/4092ccef0e8746218e05658c9408a1cec97bc909/pkg/sql/parser/placeholders.go#L169)):*
> ```Go
> 	case *Placeholder:
> 		if e, ok := v.placeholders.Value(t.Name); ok {
> 			return true, e
> ```

do you need to recurse (`true`) here?

---

*[pkg/sql/parser/placeholders.go, line 178 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-Kji8re5Ao8nP929pNZ6:-Kji8re6EDpaxYIN4WFK:bjebraa) ([raw file](https://github.com/cockroachdb/cockroach/blob/4092ccef0e8746218e05658c9408a1cec97bc909/pkg/sql/parser/placeholders.go#L178)):*
> ```Go
> 
> // replacePlaceholders replaces all placeholders in the input expression with
> // their supplied values in the planner's Placeholders map.
> ```

The comment says something about a planner but I don't see one here.

---

*[pkg/sql/parser/placeholders.go, line 180 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiIy9sV21g6XoOfmmZ:-KjiIy9sV21g6XoOfmm_:bov06nu) ([raw file](https://github.com/cockroachdb/cockroach/blob/4092ccef0e8746218e05658c9408a1cec97bc909/pkg/sql/parser/placeholders.go#L180)):*
> ```Go
> // their supplied values in the planner's Placeholders map.
> func replacePlaceholders(expr Expr, ctx *SemaContext) Expr {
> 	if ctx == nil {
> ```

if we really need to accept `nil` ctx, please document what that means.

---

*[pkg/sql/parser/placeholders.go, line 183 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiJzgDUaG8oi4a58tp:-KjiJzgDUaG8oi4a58tq:b-7ssc6u) ([raw file](https://github.com/cockroachdb/cockroach/blob/4092ccef0e8746218e05658c9408a1cec97bc909/pkg/sql/parser/placeholders.go#L183)):*
> ```Go
> 		return expr
> 	}
> 	v := &placeholdersVisitor{}
> ```

why do you need two lines to initialize a `placeholderVisitor`?

---

*[pkg/sql/parser/type_check.go, line 736 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiJPi-_okwcDo_H5ni:-KjiJPi-_okwcDo_H5nj:b-ymnlpf) ([raw file](https://github.com/cockroachdb/cockroach/blob/4092ccef0e8746218e05658c9408a1cec97bc909/pkg/sql/parser/type_check.go#L736)):*
> ```Go
> // TypeCheck implements the Expr interface.
> func (expr *Placeholder) TypeCheck(ctx *SemaContext, desired Type) (TypedExpr, error) {
> 	// Perform placeholder typing. This only happens during Prepare, when there
> ```

this comment is unclear to me. What only happens during prepare? Is this function only called during prepare?

Perhaps say that there shouldn't be any `Placeholder`s after prepare. Could we also assert that through the `SemaContext` somehow?

---

*[pkg/sql/pgwire/v3.go, line 826 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-Kjiat5BZWtPcKyTiVE4:-Kjiat5BZWtPcKyTiVE5:b-xz98my) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3023d4f47d400559e158c4b3ea9aa9477af319e/pkg/sql/pgwire/v3.go#L826)):*
> ```Go
> }
> 
> func (c *v3Conn) executeStatementParsed(
> ```

the name says "parsed" but this seems to be about "prepared" statements

---

*[pkg/sql/pgwire/v3.go, line 834 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiaQ5T2YnEKOA1lOBo:-KjiaQ5T2YnEKOA1lOBp:bsu553f) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3023d4f47d400559e158c4b3ea9aa9477af319e/pkg/sql/pgwire/v3.go#L834)):*
> ```Go
> ) error {
> 	tracing.AnnotateTrace()
> 	// Note: sql.Executor gets its Context from c.session.context, which
> ```

`sql.Executor` is a singleton; it doesn't "get" a context.

I'd remove this comment. Knowledge about how a session's ctx is used should live elsewhere I think.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-:-KjibX5BCmKxP3qgDdGj:b-yk39vj)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Thanks for the comments! I'll follow up on these.

At a high level, no, we don't keep more state in a session than before unless someone is using *named* prepared statements. The pgwire "extended protocol" does indeed prepare every statement before executing it, but it's prepared as the "default statement", which is the (single) unnamed prepared statement that is present in every session.

I think that the memory overhead here is pretty much equivalent to what it was before. One exception I can think of is if you found a way to create a statement that is much larger in its parsed form than in its text form. Then, if you were to prepare and execute this statement via the unnamed prepared statement and then leave your session open but idle, the server would keep this structure in memory pointlessly.

---

Review status: all files reviewed at latest revision, 19 unresolved discussions, all commit checks successful.

---

*[pkg/sql/executor.go, line 404 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiVEb_h6kALjcXjf9N:-KjiyXzeO7OxpI2JF1w8:b-f8sf1p) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3023d4f47d400559e158c4b3ea9aa9477af319e/pkg/sql/executor.go#L404)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I spent a while scratching my head about whether this is supposed to be a single statement or it can also be multiple ones until I saw the `switch len(stmts)` below.
Mind documenting that it needs to be a single statement?
Better yet, I think it'd be better if this took `parser.Statement` and we'd move error checking to the callers.
</blockquote></details>

Added a comment.

---

*[pkg/sql/executor.go, line 482 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiMC6aLkQu3_7oSNVT:-KjizXFevdD5qZM5duR2:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3023d4f47d400559e158c4b3ea9aa9477af319e/pkg/sql/executor.go#L482)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

nit: comment that this function needs to be directly `defer`red. Like, it can't be called from an anonymous deferred function.
</blockquote></details>

Done.

---

*[pkg/sql/executor.go, line 508 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiO7mfkDv2vdNH2CU0:-Kjiv9Wl1Sh-hLF4wIyO:bz2pwx1) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3023d4f47d400559e158c4b3ea9aa9477af319e/pkg/sql/executor.go#L508)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

are you sure this `resetForBatch()` here is correct? I don't know exactly what the consequences of it are, but the name certainly indicates that it should be done per batch of statements.
It's also called from `Executor.Prepare`, which I'm guessing is also wrong now that we support `PREPARE` as a sql statement, and not just as a wire command that always runs outside of batches.
</blockquote></details>

Yeah this is incorrect. I think it's harmless during PREPARE though.

---

*[pkg/sql/executor.go, line 509 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiOtZFPzGoYLvT6ipn:-Kjj7bp83gHvttbxI7JH:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3023d4f47d400559e158c4b3ea9aa9477af319e/pkg/sql/executor.go#L509)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

same for this `phaseTimes` - is the "startBatch" correct?
</blockquote></details>

Done.

---

*[pkg/sql/prepare.go, line 31 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiQys6zZs2jMlkJXoo:-KjimJtBn9bT8CmKb2lj:b-etbay3) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3023d4f47d400559e158c4b3ea9aa9477af319e/pkg/sql/prepare.go#L31)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Mind explaining (or just telling me) what an "empty" prepared statement represents? What does it return?
</blockquote></details>

It returns no data. This is valid behavior and happens in the wild, and we in fact have tests for it in `pgwire_test.go`.

---

*[pkg/sql/prepare.go, line 73 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiSpKhnCGkKI32uF8M:-KjimeAglhePVZOrJQpF:b-t19ybd) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3023d4f47d400559e158c4b3ea9aa9477af319e/pkg/sql/prepare.go#L73)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I don't think "logging context" is a phrase we should be using. Contexts help with more than just logging.

Feel free to ignore this, but I don't think this comment is necessary. I suspect there's no real need for the `PreparedStatements` type to hold a reference to the session; it could probably be passed in everywhere. And then you probably wouldn't have felt the need to write it, as it's more common knowledge that sessions have a context (or at least something more magic that provides a context) for all their operations. Or if we keep the session ptr, I think a comment on the `PreparedStatements` type saying that an instance is a part of a `Session` would be better,
</blockquote></details>

This was just copy pasta from `PreparedStatements.New`.

---

*[pkg/sql/prepare.go, line 74 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiYRLecbVds5ehVTIo:-KjiuprGeTayODBisrRe:bpkgcgm) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3023d4f47d400559e158c4b3ea9aa9477af319e/pkg/sql/prepare.go#L74)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Is this new method (which also comes with a duplicate comment) really necessary? Can't the caller do the parsing and call the other one?
</blockquote></details>

I think this is cleaner.

---

*[pkg/sql/prepare.go, line 95 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiZHNdRmvohmDgS3Or:-KjipxUmonks7iu51Gn9:b-gecq78) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3023d4f47d400559e158c4b3ea9aa9477af319e/pkg/sql/prepare.go#L95)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

the `queryLen` argument seems very odd to me. We moved to a nice structured type for the `statement` (really, `StatementList`, but as I was saying above I think that's a mistake), but we pass this length separately. And also the query string length is pretty silly IMHO for the memory mgmt purposes for which it's used. I suggest adding a `Size()` method to a `parser.Statement` which takes into account more than just the query length, and getting rid of this argument.
</blockquote></details>

I agree that it's odd. I agree that `Statement` should have a `Size` - in the mean time @knz suggested using query string length as a proxy.

---

*[pkg/sql/parser/placeholders.go, line 168 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiK_8hCYi6NJFAFiA1:-KjiqpEhYmMRYLutwn_r:bfnfj97) ([raw file](https://github.com/cockroachdb/cockroach/blob/4092ccef0e8746218e05658c9408a1cec97bc909/pkg/sql/parser/placeholders.go#L168)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

what if a value is not found? Should that be an error?
This probably deserves to be documented on the `placeholderVisitors` type definition.
</blockquote></details>

No - if we're in the prepare phase, we expect to see placeholders without values.

---

*[pkg/sql/parser/placeholders.go, line 169 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiKNPKDiqVAPFyjzQ8:-KjinCtPywvj5H3VojCp:bxa2pqi) ([raw file](https://github.com/cockroachdb/cockroach/blob/4092ccef0e8746218e05658c9408a1cec97bc909/pkg/sql/parser/placeholders.go#L169)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

do you need to recurse (`true`) here?
</blockquote></details>

Placeholders can be expressions, but they can't contain other placeholders. You're right.

---

*[pkg/sql/parser/placeholders.go, line 178 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-Kji8re5Ao8nP929pNZ6:-Kjir-jOBc33a9fPWvww:b7djce8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4092ccef0e8746218e05658c9408a1cec97bc909/pkg/sql/parser/placeholders.go#L178)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

The comment says something about a planner but I don't see one here.
</blockquote></details>

Yep, changed to say SemaContext.

---

*[pkg/sql/parser/placeholders.go, line 180 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiIy9sV21g6XoOfmmZ:-KjirDWD0DK9Fd1aHdb8:b-ga8zi7) ([raw file](https://github.com/cockroachdb/cockroach/blob/4092ccef0e8746218e05658c9408a1cec97bc909/pkg/sql/parser/placeholders.go#L180)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

if we really need to accept `nil` ctx, please document what that means.
</blockquote></details>

People pass `nil` SemaContexts to TypeCheck in tests a lot.

---

*[pkg/sql/parser/placeholders.go, line 183 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiJzgDUaG8oi4a58tp:-KjipA9a4aWJeKdrrUU6:b85561f) ([raw file](https://github.com/cockroachdb/cockroach/blob/4092ccef0e8746218e05658c9408a1cec97bc909/pkg/sql/parser/placeholders.go#L183)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

why do you need two lines to initialize a `placeholderVisitor`?
</blockquote></details>

I was copying how it's done in `select_name_resolution.go` without thinking about it:

```
	v := &p.nameResolutionVisitor
	*v = nameResolutionVisitor{
		err:                nil,
		sources:            sources,
		colOffsets:         make([]int, len(sources)),
		iVarHelper:         ivarHelper,
		searchPath:         p.session.SearchPath,
		foundDependentVars: false,
	}
```

Is there a reason it's done that way there?

---

*[pkg/sql/parser/type_check.go, line 736 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiJPi-_okwcDo_H5ni:-Kjisp3-6EZUP8AkQLxa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4092ccef0e8746218e05658c9408a1cec97bc909/pkg/sql/parser/type_check.go#L736)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

this comment is unclear to me. What only happens during prepare? Is this function only called during prepare?

Perhaps say that there shouldn't be any `Placeholder`s after prepare. Could we also assert that through the `SemaContext` somehow?
</blockquote></details>

Done.

---

*[pkg/sql/pgwire/v3.go, line 826 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-Kjiat5BZWtPcKyTiVE4:-KjitJuhh8x5VKsJ1sKs:bx9i3yz) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3023d4f47d400559e158c4b3ea9aa9477af319e/pkg/sql/pgwire/v3.go#L826)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

the name says "parsed" but this seems to be about "prepared" statements
</blockquote></details>

I inlined this function as it's just called in one place.

---

*[pkg/sql/pgwire/v3.go, line 834 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiaQ5T2YnEKOA1lOBo:-KjitPDq5KJVKCcpNRJF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3023d4f47d400559e158c4b3ea9aa9477af319e/pkg/sql/pgwire/v3.go#L834)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

`sql.Executor` is a singleton; it doesn't "get" a context.

I'd remove this comment. Knowledge about how a session's ctx is used should live elsewhere I think.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-:-KjikkHSY6ypfL9wgOt6:bfympli)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
I have trouble understanding from the code where the _named_ vs _unnamed_ distinction comes into play. Does unnamed mean `name == ""`? If so, maybe you can find a place to document what this means and how these unnamed statements are used by the protocol. I know this is not new code, but you seem to understand it :)

If that's true and a session has one unnamed prepared statement, consider storing the unnamed one separately from the named ones to make the code path that uses it more distinct from the paths that use cached ones, and prevent questions from the likes of me. The caching behaviour could also change I guess (you don't need to pre-parse the unnamed one if it's only executed once), but that I don't care about. I don't know if this makes sense.

More questions, sorry :) 
- is the next step to move to a cache that's not per-session? I guess many apps will prepare the same statements in all the sessions. Not to mention that we should also be caching frequent non-prepared queries.
- should we / can we store the plan that `Executor.Prepare()` creates, instead of simply the parsed statement? I expect that'd yield more performance improvements. I guess we need some invalidation story w.r.t. schema changes, but I'm already unclear how the current prepared statements are supposed to interact with schema changes that change placeholder types?

---

Review status: all files reviewed at latest revision, 19 unresolved discussions, all commit checks successful.

---

*[pkg/sql/prepare.go, line 73 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiSpKhnCGkKI32uF8M:-KjjERfwf16oPmtvlOBd:b-bcl892) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3023d4f47d400559e158c4b3ea9aa9477af319e/pkg/sql/prepare.go#L73)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

This was just copy pasta from `PreparedStatements.New`.
</blockquote></details>

I realize that, I think it's bad everywhere. And a duplicate comment is also bad for other reasons; you should have one of the comments redirect to the other.
Or I would get rid of the method :)

What bothers me about this method also is the name `New`. I had to read it three times to understand what type it's defined on and what it returns. And now there's two of them.

---

*[pkg/sql/prepare.go, line 95 at r12](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiZHNdRmvohmDgS3Or:-KjjF75LgyKwHO3_hx67:b-mcwh1l) ([raw file](https://github.com/cockroachdb/cockroach/blob/c3023d4f47d400559e158c4b3ea9aa9477af319e/pkg/sql/prepare.go#L95)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I agree that it's odd. I agree that `Statement` should have a `Size` - in the mean time @knz suggested using query string length as a proxy.
</blockquote></details>

using whatever as a proxy is fine, but doing so by separately plumbing the length is what's bothering me. Consider making it a member even if you still only count the original size. It'd give you a place to hang a TODO from, and it'd hide the dirty laundry. 

---

*[pkg/sql/parser/placeholders.go, line 168 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiK_8hCYi6NJFAFiA1:-KjjFaDLYcSO35BJ5419:b-pfygki) ([raw file](https://github.com/cockroachdb/cockroach/blob/4092ccef0e8746218e05658c9408a1cec97bc909/pkg/sql/parser/placeholders.go#L168)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

No - if we're in the prepare phase, we expect to see placeholders without values.
</blockquote></details>

but why do we use this visitor in the prepare phase if it's not expected to do anything?

---

*[pkg/sql/parser/placeholders.go, line 183 at r9](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-KjiJzgDUaG8oi4a58tp:-KjjGQGowaIY5MOP94Xr:bx173h8) ([raw file](https://github.com/cockroachdb/cockroach/blob/4092ccef0e8746218e05658c9408a1cec97bc909/pkg/sql/parser/placeholders.go#L183)):*
<details><summary><i>Previously, jordanlewis (Jordan Lewis) wrote…</i></summary><blockquote>

I was copying how it's done in `select_name_resolution.go` without thinking about it:

```
	v := &p.nameResolutionVisitor
	*v = nameResolutionVisitor{
		err:                nil,
		sources:            sources,
		colOffsets:         make([]int, len(sources)),
		iVarHelper:         ivarHelper,
		searchPath:         p.session.SearchPath,
		foundDependentVars: false,
	}
```

Is there a reason it's done that way there?
</blockquote></details>

that one uses `&p.nameResolutionVisitor` to avoid an allocation (as documented at that field's definition). You may want to preallocate your visitor in a similar way

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-:-KjjHdQQIo-3kcciY4tu:bbio8w7)*
<!-- Sent from Reviewable.io -->

jordanlewis commented:
Yes, unnamed means `name == ""`.  It's already thinly documented. I'll improve it a little, but the behavior of the unnamed prepared query and named prepared queries is exactly identical except for the fact that the unnamed prepared query can be overwritten, so I don't know if it would be a good idea to separate the code paths. In other words, clients can in fact reuse the unnamed prepared statement just as they can named prepared statements.

Caching plans is something that I'm planning to work on for 1.1. Like you mention, we do need an invalidation mechanism - naively, the plans could store a schema version and always get invalidated if the current version is different from the stored version. Presently, prepared statements get re-type-checked upon execution, so schema changes that invalidate them will fail at execute time, which is what Postgres does. I think it's desirable behavior. Executing a prepared statement should be equivalent to running the analogous non-prepared statement.

I haven't thought too much about expanding the scope of statement caching, or frequent statement caching. I'm interested in working on that though at some point and I think it could certainly work.

---

Review status: all files reviewed at latest revision, 19 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/15639#-:-Kjn4gl7zD0wPgRXhZbS:b-a0q00v)*
<!-- Sent from Reviewable.io -->
